[TASK] add global statistics for individual sites

belzebub40k opened the pull request:
As suggested in #98 this adds global statistics per configured site.

The sites for which individual statistics are created can be specified in the config and if the list is empty only the already existing global statistic is created.

```
[respondd]
sites = ["ffwi","ffmz"]
```

Site specific and global statistics are stored in different measurements. For _influxdb_ site specific data is stored inside `(global|firmware|model|autoupdater)_site` with tag `site=$sitecode`. As _graphite_ doesn't support tags (yet) one measurement `(global|firmware|model)_$sitecode` is created for each configured site.

Example output could be viewed under https://stats.freifunk-mwu.de/dashboard/db/freifunk-mwu-test

genofire commented on runtime/config.go:
```diff
@@ -12,6 +12,7 @@ type Config struct {
 		Enable          bool     `toml:"enable"`
 		Synchronize     Duration `toml:"synchronize"`
 		Interfaces      []string `toml:"interfaces"`
+		Sites		[]string `toml:"sites"`
```
need spaces, use the gofmt linter

genofire commented on runtime/stats.go:
```diff
@@ -20,34 +20,43 @@ type GlobalStats struct {
+
+	var nodelist []*Node
+
+	if site != "global" {
```
define `"global"` as const

genofire commented on runtime/stats.go:
```diff
@@ -20,34 +20,43 @@ type GlobalStats struct {
+		nodelist = nodes.Select(func(n *Node) bool {
+			return n.Online })
+	}
+
```
`nodes.RLock()` and `nodes.RUnlock()` not needed anymore -> is locked in `Select` already

belzebub40k commented:
I fixed the things you mentioned.

What about creating a map of `GlobalStats struct` inside `NewGlobalStats()` instead of changing the struct?

belzebub40k commented:
Updates PR to iterate over all nodes only once and save the result in `map[string]*GlobalStats`.

genofire commented on cmd/query.go:
```diff
@@ -26,7 +26,7 @@ var queryCmd = &cobra.Command{
 		nodes := runtime.NewNodes(&runtime.Config{})
 
-		collector := respond.NewCollector(nil, nodes, []string{iface}, 0)
+		collector := respond.NewCollector(nil, nodes, []string{}, []string{iface}, 0)
```
Use the global site.

belzebub40k replied:
The global site is always initialized. []string{} is just the list of additional sites and not needed for `query`.

genofire commented on respond/collector.go:
```diff
@@ -15,6 +15,8 @@ import (
 	"github.com/FreifunkBremen/yanic/runtime"
 )
 
+const GLOBAL = "global"
```
Rename the const to SITE_GLOBAL

belzebub40k replied:
This `const` isn't used. Will remove it.

genofire commented on runtime/stats.go:
```diff
@@ -1,6 +1,7 @@
 package runtime
 
 const DISABLED_AUTOUPDATER = "disabled"
+const GLOBAL = "global"
```
Only one const ... see collector

belzebub40k replied:
See previous comment.

belzebub40k commented:
> Looks nearly good for me, now the tests?

Will start to write the tests now.

corny commented:
Global and site specific data points should be stored in different measurements.

genofire commented:
Stimmt, graphite Mann ja nicht mit tags umgehen

belzebub40k commented:
> Global and site specific data points should be stored in different measurements.

I would prefer to have only one measurement for all in influxdb as it makes the handling in grafana much easier as you can use one dashboard for global and site-specific data. But if your are really unhappy with this solution my suggestion would be to leave the _global_ measurement as it is and store all site-specific data in _global_site_ with tag `site=site_code`. As graphite doesn't support tags we have to use one measurement per site.

belzebub40k commented:
I experienced a problem when counting the gateways per site because `SiteCode` is empty sometimes.

Below is a dump I created with `spew.Dump(node)`
```
(*runtime.Node)(0xc420144d20)({
 Address: (*net.UDPAddr)(<nil>),
 Firstseen: (jsontime.Time) {
  time: (time.Time) 2015-05-01 17:23:59 +0100 +0100
 },
 Lastseen: (jsontime.Time) {
  time: (time.Time) 2017-11-26 15:05:12 +0100 CET
 },
 Online: (bool) true,
 Statistics: (*data.Statistics)(0xc42012cd20)({
  NodeID: (string) (len=12) "02410a250007",
  Clients: (data.Clients) {
   Wifi: (uint32) 0,
   Wifi24: (uint32) 0,
   Wifi5: (uint32) 0,
   Total: (uint32) 0
  },
  RootFsUsage: (float64) 0,
  LoadAverage: (float64) 2.75,
  Memory: (data.Memory) {
   Cached: (uint32) 1811948,
   Total: (uint32) 3015588,
   Buffers: (uint32) 195532,
   Free: (uint32) 132464
  },
  Uptime: (float64) 3.8913005e+06,
  Idletime: (float64) 1.683778543e+07,
  GatewayIPv4: (string) "",
  GatewayIPv6: (string) "",
  GatewayNexthop: (string) "",
  Processes: (struct { Total uint32 "json:\"total\""; Running uint32 "json:\"running\"" }) {
   Total: (uint32) 248,
   Running: (uint32) 3
  },
  MeshVPN: (*data.MeshVPN)(<nil>),
  Traffic: (struct { Tx *data.Traffic "json:\"tx\""; Rx *data.Traffic "json:\"rx\""; Forward *data.Traffic "json:\"forward\""; MgmtTx *data.Traffic "json:\"mgmt_tx\""; MgmtRx *data.Traffic "json:\"mgmt_rx\"" }) {
   Tx: (*data.Traffic)(0xc42015ede0)({
    Bytes: (float64) 1.152308749661e+13,
    Packets: (float64) 1.0379155709e+10,
    Dropped: (float64) 2.300964e+06
   }),
   Rx: (*data.Traffic)(0xc42015ee00)({
    Bytes: (float64) 1.163723502662e+12,
    Packets: (float64) 7.073107282e+09,
    Dropped: (float64) 0
   }),
   Forward: (*data.Traffic)(0xc42015ee20)({
    Bytes: (float64) 3.233804231637e+12,
    Packets: (float64) 2.2566065194e+10,
    Dropped: (float64) 0
   }),
   MgmtTx: (*data.Traffic)(0xc42015ee40)({
    Bytes: (float64) 5.4600307316e+10,
    Packets: (float64) 1.67716696e+08,
    Dropped: (float64) 0
   }),
   MgmtRx: (*data.Traffic)(0xc42015ee60)({
    Bytes: (float64) 3.411465894664e+12,
    Packets: (float64) 8.472243523e+09,
    Dropped: (float64) 0
   })
  },
  Switch: (map[string]*data.SwitchPort) <nil>,
  Wireless: (data.WirelessStatistics) <nil>
 }),
 Nodeinfo: (*data.NodeInfo)(0xc42015c900)({
  NodeID: (string) (len=12) "02410a250007",
  Network: (data.Network) {
   Mac: (string) (len=17) "02:41:0a:25:00:07",
   Addresses: ([]string) (len=4 cap=4) {
    (string) (len=19) "fe80::42:aff:fe25:7",
    (string) (len=21) "fd37:b4dc:4b1e::a25:7",
    (string) (len=20) "2a03:2260:11a::a25:7",
    (string) (len=25) "fe80::84cd:deff:feb1:8bcc"
   },
   Mesh: (map[string]*data.BatInterface) (len=1) {
    (string) (len=5) "mzBAT": (*data.BatInterface)(0xc420073090)({
     Interfaces: (struct { Wireless []string "json:\"wireless,omitempty\""; Other []string "json:\"other,omitempty\""; Tunnel []string "json:\"tunnel,omitempty\"" }) {
      Wireless: ([]string) <nil>,
      Other: ([]string) <nil>,
      Tunnel: ([]string) (len=2 cap=4) {
       (string) (len=17) "02:00:0a:25:00:07",
       (string) (len=17) "02:00:0a:25:01:07"
      }
     }
    })
   },
   MeshInterfaces: ([]string) <nil>
  },
  Owner: (*data.Owner)(<nil>),
  System: (data.System) {
   SiteCode: (string) ""
  },
  Hostname: (string) (len=6) "spinat",
  Location: (*data.Location)(<nil>),
  Software: (data.Software) {
   Autoupdater: (struct { Enabled bool "json:\"enabled,omitempty\""; Branch string "json:\"branch,omitempty\"" }) {
    Enabled: (bool) false,
    Branch: (string) ""
   },
   BatmanAdv: (struct { Version string "json:\"version,omitempty\""; Compat int "json:\"compat,omitempty\"" }) {
    Version: (string) (len=6) "2017.1",
    Compat: (int) 0
   },
   Fastd: (struct { Enabled bool "json:\"enabled,omitempty\""; Version string "json:\"version,omitempty\"" }) {
    Enabled: (bool) true,
    Version: (string) (len=3) "v17"
   },
   Firmware: (struct { Base string "json:\"base,omitempty\""; Release string "json:\"release,omitempty\"" }) {
    Base: (string) (len=6) "Ubuntu",
    Release: (string) (len=5) "14.04"
   },
   StatusPage: (struct { API int "json:\"api\"" }) {
    API: (int) 0
   }
  },
  Hardware: (data.Hardware) {
   Nproc: (int) 6,
   Model: (string) ""
  },
  VPN: (bool) true,
  Wireless: (*data.Wireless)(<nil>)
 }),
 Neighbours: (*data.Neighbours)(<nil>)
})

```

genofire commented on database/all/internal.go:
```diff
@@ -43,9 +43,9 @@ func (conn *Connection) InsertLink(link *runtime.Link, time time.Time) {
 }
 
-func (conn *Connection) InsertGlobals(stats *runtime.GlobalStats, time time.Time) {
+func (conn *Connection) InsertGlobals(stats *runtime.GlobalStats, time time.Time, site string) {
```
maybe use of `stats *rmap[string]runtime.GlobalStats` and handle it inside of `InsertGlobals`.
So we could use in graphite `stats["global"]` to fix the multiple inserts in graphite.

Alternative:
catch in graphite every site which is not "global" before it is run.
(just a `if site==SITE_GLOBAL {return}` at the beginning)

belzebub40k replied:
Yes, _graphite_ is still broken in the branch. I'll try to work on this today and go for the `(global|firmware|model)_$site` schema.

genofire replied:
But keep`(global|firmware|model)` if `$sitecode==global` ?

belzebub40k replied:
Yes, `$sitecode==global` will go into the existing measurements.

belzebub40k commented:
Ok, I think I'm done and this is ready to be reviewed.

All databases _(logging, influxdb and graphite)_ are supported. Current global stats are untouched for both influxdb and graphite. Site-specific global stats are saved either in `(global|firmware|model)_$sitecode` _(graphite)_ or `(global|firmware|model|autoupdater)_site` with tag `site=$sitecode` _(influxdb)_.

genofire commented on database/influxdb/global.go:
```diff
@@ -8,11 +8,31 @@ import (
+	var (
+		tags models.Tags
+
+		measurementGlobal = MeasurementGlobal
```
use `measurementGlobal := MeasurementGlobal` instatt of `var measurementGlobal = MeasurementGlobal`

belzebub40k replied:
Done.

genofire commented on database/graphite/global.go:
```diff
@@ -7,20 +7,32 @@ import (
-	c.addCounterMap(CounterMeasurementFirmware, stats.Firmwares, time)
+func (c *Connection) InsertGlobals(stats *runtime.GlobalStats, time time.Time, site string) {
+	var (
+		measurementGlobal = MeasurementGlobal
```
use `x := y` instatt of `var x = y`

belzebub40k replied:
Done.

genofire commented on database/influxdb/global_test.go:
```diff
@@ -9,14 +9,22 @@ import (
 
-	// check fields
+	// check SITE_GLOBAL fields
+	fields = GlobalStatsFields(stats[SITE_GLOBAL])
```
...

belzebub40k replied:
Done.

genofire commented on runtime/stats_test.go:
```diff
@@ -8,26 +8,50 @@ import (
 	"github.com/FreifunkBremen/yanic/data"
 )
 
+const SITE_GLOBAL = "global"
```
use `GLOBAL_SITE` instatt ... (it is already definied in this package)

belzebub40k replied:
Done.

genofire commented on database/graphite/global.go:
```diff
@@ -7,20 +7,32 @@ import (
+		counterMeasurementFirmware = CounterMeasurementFirmware
+	)
+
+	if site != "global" {
```
use `runtime.GLOBAL_SITE`

belzebub40k replied:
Done.

genofire commented on database/influxdb/global.go:
```diff
@@ -8,11 +8,31 @@ import (
+		counterMeasurementAutoupdater = CounterMeasurementAutoupdater
+	)
+
+	if site != "global" {
```
use `runtime.GLOBAL_SITE`

belzebub40k replied:
Done.

genofire commented:
Danke für die ganze Mühe.

belzebub40k commented:
> Danke für die ganze Mühe.

Gerne und danke für das konstruktive Feedback. Da ich vorher noch nicht wirklich mit Go gearbeitet habe konnte ich beim Implementieren einiges über Go gelernt.

genofire commented on runtime/stats.go:
```diff
@@ -20,41 +23,62 @@ type GlobalStats struct {
-				}
+			result[GLOBAL_SITE].Add(node)
+			
+			site := node.Nodeinfo.System.SiteCode
```
`node.Nodeinfo` is a pointer and could be `nil`. It would crash ...

surround it with a `if` (see line 71):
https://github.com/FreifunkBremen/yanic/pull/102/files#diff-913c82e313e28317b4cb3a41fabc87b8R71

genofire commented on database/graphite/global.go:
```diff
@@ -7,20 +7,30 @@ import (
+
+	c.addPoint(GlobalStatsFields(measurementGlobal, stats))
+	c.addCounterMap(counterMeasurementModel, stats.Models, time)
+	c.addCounterMap(counterMeasurementFirmware, stats.Firmwares, time)
```
could you fix something for me? (if you works on it)
Add the `CounterMap` of `stats.Autoupdater`

genofire replied:
`c.addCounterMap(counterMeasurementAutoupdater, stats.Autoupdater, time)`

belzebub40k replied:
Shouldn't work on two things at the same time. Fixed.

genofire commented on runtime/stats.go:
```diff
@@ -20,33 +23,32 @@ type GlobalStats struct {
-					result.Autoupdater.Increment(info.Software.Autoupdater.Branch)
-				} else {
-					result.Autoupdater.Increment(DISABLED_AUTOUPDATER)
+				site := node.Nodeinfo.System.SiteCode
```
`site := info.System.SiteCode`

belzebub40k replied:
Done.

genofire commented on database/graphite/global.go:
```diff
@@ -11,16 +11,19 @@ func (c *Connection) InsertGlobals(stats *runtime.GlobalStats, time time.Time, s
 	c.addPoint(GlobalStatsFields(measurementGlobal, stats))
 	c.addCounterMap(counterMeasurementModel, stats.Models, time)
 	c.addCounterMap(counterMeasurementFirmware, stats.Firmwares, time)
+	c.addCounterMap(counterMeasurementAutoupdater, stats.Firmwares, time)
```
`stats.Autoupdater` xD

belzebub40k replied:
Damn copy and paste >.<
